docs_src/design/adr/security/0017-consul-security.md

```diff
@@ -0,0 +1,111 @@
+
+Since configuration data can affect the runtime behavior of services,
+compensating controls must be introduced in order to mitigate the risks introduced
+by moving configuration from a static file into to an HTTP-accessible service with mutable state.
+
+The current practice is that Consul is exposed via unencrypted HTTP in anonymous read/write mode
+to all processes and EdgeX services running on the host machine.
+
+## Decision
+
+Consul will be configured with access control list (ACL) functionality enabled,
+and each EdgeX service will utilize a Consul access token to authenticate to Consul.
+Consul access tokens will be requested from the Vault Consul secrets engine
+(to avoid introducing additional bootstrapping secrets).
+
+DNS will be disabled via configuration as it is not used in EdgeX.
+
+**Consul Access Via API Gateway**
+
+In security enabled EdgeX, the API gateway will be configured to
+proxy the Consul service over the `/consul` path,
+using the `request-transformer` plugin
+to add the global management token to incoming requests
+via the `X-Consul-Token` HTTP header.
+Thus, ability to access remote APIs also grants the ability
+to modify Consul's key-value store.
+
+Proxying of the Consul UI is problematic and there is no current solution,
+which would involve proper balacing of the externally-visible URL,
+the path-stripping effect (or not) of the proxy,
+Consul's `ui_content_path`,
+and UI authentication
+(the `request-transfomer` does not work on the UI).
+
+
+## Consequences
+
+Full implementation of this ADR will deny Consul access to all existing Consul clients.
+To limit the impacts of the change, deployment will take place in phases:
+
+### Phase 1
+
+- Vault bootstrapper will install Vault Consul secrets engine.
+- Secretstore-setup will create a Vault token for consul secrets engine configuration.
+- Consul will be started with Consul ACLs enabled with persistent agent tokens and a default "allow" policy.
+- Consul bootstrapper will create a bootstrap management token
+  and use the provided Vault token to (re)configure the Consul secrets engine in Vault.
+- Consul bootstrapper will create an agent ("node") policy and token and install it into the agent.
+- Create a global management token (stored in a file) for use in remote configuration.
+- (Docker-only) Open a port to signal that Consul bootstrapping is completed.
+  (Integrate with `ready_to_run` signal.)
+
+### Phase 2
+
+- Consul bootstrapper will install a role in Vault that creates global-management tokens in Consul with no TTL.
+- Registry and configuration client libraries will be modified to accept a Consul access token.
+- go-mod-bootstrap will have contain the necessary glue logic to
+  request a service-specifc Consul access token from Vault
+  and persist it as a service-specific secret.
+- Consul configuration will be changed to a default "deny" policy
+  once all services have been changed to authenticated access mode.
```

Comment:
the original defined Consul's ACL policies may need to rectified as the meaning of "allow" versus "deny" now has the opposite interpretation in terms of access control.  Reference: https://www.consul.io/docs/agent/options#acl_default_policy. 
```
In "allow" mode, ACLs are a denylist: any operation not specifically prohibited is allowed. 
In "deny" mode, ACLs are an allowlist: any operation not specifically allowed is blocked. 
```

Reply:
assign global mgmt policy to agent token can resolve this situation and thus we can switch to stricter policy once "deny" mode is turned-on.

Reply:
Change has been made.

Reply:
issue resolved